docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,673 @@
+
+:::info transformRequest naming
+We are using `transformRequest(url)` and `warmupRequest(url)` in the current version of this proposal so it is easier to discuss and understand for users used to Vite's current API. Before releasing, we can take the opportunity to review these names too. For example, it could be named `environment.processModule(url)` or `environment.loadModule(url)` taking a page from Rollup's `context.load(id)` in plugin hooks. For the moment, we think keeping the current names and delaying this discussion is better.
+:::
+
+For the default SSR environment, Vite creates a module runner that implements evaluation using `new AsyncFunction` running in the same runtime as the server. This runner is an instance of `ModuleRunner` that exposes:
+
+```ts
+class ModuleRunner {
+  /**
+   * URL to execute. Accepts file path, server path, or id relative to the root.
+   * Returns an instantiated module (same as in ssrLoadModule)
+   */
+  public async executeUrl(url: string): Promise<Record<string, any>>
+  /**
+   * Entry point URL to execute. Accepts file path, server path or id relative to the root.
+   * In the case of a full reload triggered by HMR, this is the module that will be reloaded.
+   * If this method is called multiple times, all entry points will be reloaded one at a time.
+   * Returns an instantiated module
+   */
+  public async executeEntrypoint(url: string): Promise<Record<string, any>>
+  /**
+   * Other ModuleRunner methods...
+   */
+```
+
+The SSR module runner instance is accessible through `server.ssrModuleRunner`. It isn't part of its associated environment instance because, as we explained before, other environments' module runners could live in a different runtime (for example for the browser, a module runner in a worker thread as used in Vitest, or an edge runtime like workerd). The communication between `server.ssrModuleRunner` and `server.environment('ssr')` is implemented through direct function calls. Given a Vite server configured in middleware mode as described by the [SSR setup guide](#setting-up-the-dev-server), let's implement the SSR middleware using the environment API. Error handling is omitted.
+
+```js
+app.use('*', async (req, res, next) => {
+  const url = req.originalUrl
+
+  // 1. Read index.html
+  let template = fs.readFileSync(path.resolve(__dirname, 'index.html'), 'utf-8')
+
+  // 2. Apply Vite HTML transforms. This injects the Vite HMR client,
+  //    and also applies HTML transforms from Vite plugins, e.g. global
+  //    preambles from @vitejs/plugin-react
+  template = await server.transformIndexHtml(url, template)
+
+  // 3. Load the server entry. executeEntryPoint(url) automatically transforms
+  //    ESM source code to be usable in Node.js! There is no bundling
+  //    required, and provides full HMR support.
+  const { render } = await server.ssrModuleRunner.executeEntryPoint(
+    '/src/entry-server.js',
+  )
```

Comment:
If a user wants to run a code in workerd, would the code be like below?
```js
// the input and the output depends on the environment
const response = await server.environment('workerd').run(nodeReqToNativeReq(req))
const html = getResponseContent(response)
```

Reply:
Let's merge this thread into https://github.com/vitejs/vite/pull/16089#discussion_r1512147458